### PR TITLE
Enabled successor command/checks to extend testing capabilities

### PR DESCRIPTION
**Pull request by sifraser**

### CODE REVIEW

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        private val Log = LoggerFactory.getLogger(VerifiableScenarioCursor::class.java)
+    }
+}
+
+sealed interface SystemCursor<AF : ActionFactory, CF : CheckFactory> {
+    fun apply(iteration: SystemIteration): System<AF, CF>
+    fun destroy() {}
+}
+
+class MutableSystemCursor<AF : ActionFactory, CF : CheckFactory>(val system: MutableSystem<AF, CF>) :
```
Comment (sifraser) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt]: We should almost certainly apply the same cursor/mutable system pattern for action generators

File: tictactoe/eacs/build.gradle [hunk 2]
```diff
@@ -112,3 +117,6 @@ task rollForwardTest(type: Test) {
         "com.anaplan.engineering.azuki.persistence.verification.instance.jar": sampleImplementationV2ClasspathJar.archiveFile.get().asFile.absolutePath,
     ]
 }
+
+check.dependsOn(rollBackTest)
+check.dependsOn(rollForwardTest)
```
Comment (sifraser) [inline tictactoe/eacs/build.gradle:122]: The default impls and these dependencies are to ensure all tests are run on full build -- noticed they weren't running and so I had unexpected confidence in change.

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt [hunk 3]
```diff
@@ -219,7 +218,7 @@ class JUnitScenarioRunner<
                                 scenario,
                                 scenarioName)
                         when (verifiableScenarioRunner.run()) {
-                            VerifiableScenarioRunner.Result.UnsupportedAction -> unsupported("Skipping - unsupported action found")
+                            VerifiableScenarioRunner.Result.UnsupportedCommand -> unsupported("Skipping - unsupported action found")
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/JUnitScenarioRunner.kt:221]: ⛏️ I presume you'd want to update the logged message as well.

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 4]
```diff
@@ -209,10 +218,12 @@ class MultiOracleScenarioRunner<
...
-            val systemFactory = implementation.createSystemFactory()
-            val declarations = scenario.definitions(systemFactory.actionFactory)
-            val buildActions = scenario.buildActions(systemFactory.actionFactory)
-            if (UnsupportedAction in declarations || UnsupportedAction in buildActions) {
+            val systemFactory =
+                implementation.createSystemFactory() as? QueryableSystemFactory<AF, CF, QF, AGF, *, QueryableSystem<AF, CF>>
+                    ?: throw IllegalStateException("Trying to query, but system factory does not create queryable systems")
+            val declarations = scenario.declarations(systemFactory.actionFactory)
+            val commands = scenario.commands(systemFactory.actionFactory)
+            if (UnsupportedAction in declarations || UnsupportedAction in commands) {
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:226]: ❓ Is it supposed to stay as `UnsupportedAction` or should it be `UnsupportedCommand`?

**Comment by sifraser** [comment]
Spotted an issue .. converted to draft

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
+package com.anaplan.engineering.azuki.core.scenario
+
+import com.anaplan.engineering.azuki.core.system.*
+import org.slf4j.LoggerFactory
+
+class VerifiableScenarioCursor<AF : ActionFactory, CF : CheckFactory, S : VerifiableSystem<AF, CF>, SF : VerifiableSystemFactory<AF, CF, *, *, *, S>>(
+    private val systemFactory: SF,
+    private val scenario: VerifiableScenario<AF, CF>,
+) {
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt:9]: ❓ Would it make sense to implement `kotlin.collections.Iterator`, given that this appears to implement an iterator interface?

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        if (system is MutableSystem<*, *>) {
+            system.applyIteration(iteration)
+            system
+        } else {
+            systemDefinition = systemDefinition.copy(
+                commands = systemDefinition.commands + iteration.commands,
+                checks = iteration.checks,
+                regardlessOfActions = iteration.regardlessOfActions
+            )
+            systemFactory.create(systemDefinition)
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt:90]: ❓ Given that the cursor keeps track of the current state of the system, wouldn't it make sense to replace the value of `system` with this?

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/SystemFactory.kt [hunk 7]
```diff
@@ -58,13 +67,33 @@ interface PersistableSystem {
...
+    val regardlessOfActions: List<List<Action>> = emptyList(),
+    val queries: List<Query<*>> = emptyList(),
+    val forAllQueries: List<DerivedQuery<*>> = emptyList(),
+    val actionGenerators: List<ActionGenerator> = emptyList(),
+)
+
 sealed class VerificationResult {
     // The system was valid, the checks were run, and they were satisfied
-    class Verified(
+    open class Verified(
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/SystemFactory.kt:92]: ❓ Why does this need to be open?

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/SystemFactory.kt [hunk 8]
```diff
@@ -58,13 +67,33 @@ interface PersistableSystem {
...
         // val coverage
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/SystemFactory.kt:93]: ⛏️ Might as well get rid of this...

**Review by donbex — APPROVED** [review]

**Pull request merged**